### PR TITLE
Corridor delineation based on river valley

### PR DESCRIPTION
**Pull request by fnattino**
I have tested the river corridor delineation using the valley boundary as initial guess of the corridor. 

Considering the length of the river corridor notebook for Bucharest, I have opted for not continuing in the same document. For ease of testing, I have also split the workflow in two parts and tried to clean up the code we have so far. The result is two notebooks:
* a notebook to download all the required OSM data (note that all retrieved data is also checked in this branch)
* a notebook where the preprocessing and delineation is carried out.

Can you have a look @cforgaci ? I have included a "repulsive" term in the weights to "push" the corridor edges outside the valley region. The choice of the values for the coefficient is pretty arbitrary at the moment, but the results seems relatively good to me?

### CODE REVIEW

File: notebooks/corridor/corridor_bucharest_river-valley.qmd [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+## 3. Corridor edge delineation ----
+
+We carry out the delineation of the river corridor in few steps.
+
+First, we consider the nodes of the network within the area of interest (here defined using the bounding box of the city) and label them according to which side of the river they fall on:
+
+```{r}
+# split polygon in two fragments using the waterway geometry as dividing line
+# if more than two fragments are obtained, keep only the two largest fragments
```
Comment (cforgaci) [inline notebooks/corridor/corridor_bucharest_river-valley.qmd:141]: Does this mean that any islands created by the multiple branches of the river are removed?

File: notebooks/corridor/corridor_bucharest_river-valley.qmd [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    st_as_sfc()
+
+# split the area of interest in the two river sides
+river_sides <- split_river_sides(bbox_polygon, waterway)
+
+# label nodes with side of the river they fall on
+network_classified <- network_cleaned |>
+    activate("nodes") |>
+    mutate(side_1 = node_intersects(river_sides[1,])) |>
+    mutate(side_2 = node_intersects(river_sides[2,]))
```
Comment (cforgaci) [inline notebooks/corridor/corridor_bucharest_river-valley.qmd:171]: If `side_1` and `side_2` are exclusive, then this should work too:
```suggestion
    mutate(side_1 = node_intersects(river_sides[1,]),
           side_2 = !side_1)
```
The advantage is that you do not run `node_interesects()` again. You can also put both in the same `mutate()` call.

File: notebooks/corridor/corridor_bucharest_river-valley.qmd [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+not_intersects <- function(x, y) !st_intersects(x, y)
+
+network_split <- network_weights  |>
+    activate("edges") |>
+    st_filter(waterway, .predicate = not_intersects) |>
+    activate("nodes") |>
+    filter(!node_is_isolated())
+```
+
+TODO: is enough to drop isolated nodes, or should we check for the largest group components?
```
Comment (cforgaci) [inline notebooks/corridor/corridor_bucharest_river-valley.qmd:261]: Indeed checking for the largest component should be more reliable as the removal of crossing segment may create more then two connected components besides isolated nodes.

**Review by cforgaci — COMMENTED** [review]
Thanks! The valley edge detection works quite well with the penalty cost you introduced! I think there is a trade-off between the penalty and the way the valley was detected in the first place. With this penalty we assume that the correct delineation will always be outside the detected valley edge; if the valley edge is "overestimated" in the first place, then we might need to account for delineations that are still within it. What are your thoughts on this?

I left two small comments.

File: notebooks/corridor/corridor_bucharest_river-valley.qmd (hunk 2 above)
Comment (cforgaci) [reply notebooks/corridor/corridor_bucharest_river-valley.qmd:171]: This does not work as the bbox considered for the nodes is different from the extended bbox used for the edges.

**Review by cforgaci — APPROVED** [review]
👍

File: notebooks/corridor/corridor_bucharest_river-valley.qmd (hunk 1 above)
Comment (fnattino) [reply notebooks/corridor/corridor_bucharest_river-valley.qmd:141]: As discussed, yes, smaller regions like islands will be removed. But "artificial" fragments created e.g. by the river flowing close to the boundary of the polygon would be ignored.

File: notebooks/corridor/corridor_bucharest_river-valley.qmd (hunk 3 above)
Comment (fnattino) [reply notebooks/corridor/corridor_bucharest_river-valley.qmd:261]: Good point, we should definitely keep the two largest components in the final version in the package!

**Comment by fnattino** [comment]
> I think there is a trade-off between the penalty and the way the valley was detected in the first place. With this penalty we assume that the correct delineation will always be outside the detected valley edge; if the valley edge is "overestimated" in the first place, then we might need to account for delineations that are still within it. What are your thoughts on this?

Totally agree, we will find out whether the values chosen here create a sound balance by testing on other cities..

**Pull request merged**